### PR TITLE
Set up the test suite for shallowUntilTarget rendering

### LINKED ISSUE #15: Set up the test suite for shallowUntilTarget rendering
Once we have https://github.com/mozilla/addons-code-manager/issues/8 , set the test suite up for shallow rendering. Since we'll be using Redux, we will need a strategy like [shallowUntilTarget](https://github.com/mozilla/addons-frontend/search?q=shallowUntilTarget&unscoped_q=shallowUntilTarget).

Here is a write-up of what worked well in the addons-frontend test suite: https://hacks.mozilla.org/2018/04/testing-strategies-for-react-and-redux/

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #15 

I had some trouble with types, so there will probably be some comments and suggestions around those. In addition to that I changed the signature so it accepts an object instead of individual arguments. I think it's cleaner, and that's what we tend to do most of the time, but it is a bit more of a hassle to call.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -13,7 +13,7 @@
     // https://github.com/typescript-eslint/typescript-eslint/blob/5b0b3d9edbcb3ab588a34c431037d9deece30824/packages/eslint-plugin/docs/rules/indent.md#options
     "@typescript-eslint/indent": ["error", 2],
     "import/no-extraneous-dependencies": ["error", {
-        "devDependencies": ["**/*.spec.*", "src/setupTests.tsx"]
+        "devDependencies": ["**/*.spec.*", "src/setupTests.tsx", "src/test-helpers.tsx"]
```
Comment (willdurand) [inline .eslintrc]: Why is this needed?

File: src/test-helpers.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+import { shallowUntilTarget } from './test-helpers';
+
+describe(__filename, () => {
+  describe('shallowUntilTarget', () => {
+    function ExampleBase() {
+      return <div>Example component</div>;
+    }
+
+    function wrapper() {
```
Comment (willdurand) [inline src/test-helpers.spec.tsx]: We could probably switch to arrow functions in this file.

File: src/test-helpers.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
+/* eslint-disable react/no-multi-comp */
```
Comment (willdurand) [inline src/test-helpers.spec.tsx]: Why is this rule disabled?

File: .eslintrc (hunk 1 above)
Comment (bobsilverberg) [reply .eslintrc]: Without it I get "2:1   error    'enzyme' should be listed in the project's dependencies, not devDependencies" on `/src/test-helpers.tsx`. I assume this is  because `ESLint` thinks this file is a prod file and not a test-only file, and this fixes that.

File: package.json [hunk 4]
```diff
@@ -3,6 +3,7 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@types/common-tags": "1.8.0",
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply package.json]: Sure, it's just used for an error message but I can probably do without it.

**Comment by bobsilverberg** [comment]
Thanks for the review @willdurand. I've addressed the comments and pushed a new commit.

File: src/components/App/index.spec.tsx [hunk 5]
```diff
@@ -19,11 +19,11 @@ describe(__filename, () => {
...
   }: RenderParams = {}) => {
-    // TODO: Use shallowUntilTarget()
-    // https://github.com/mozilla/addons-code-manager/issues/15
-    const root = shallow(<App authToken={authToken} />, {
-      context: { store },
-    }).shallow();
+    const root = shallowUntilTarget({
+      componentInstance: <App authToken={authToken} />,
+      targetComponent: AppBase,
+      options: { shallowOptions: { context: { store } } },
```
Comment (willdurand) [inline src/components/App/index.spec.tsx]: This is interesting (not related to your patch): we set the `store` to the `context` whereas in addons-frontend, it was injected as a prop, which caused issues with the react-redux v6.

File: src/test-helpers.tsx [hunk 6]
```diff
@@ -25,3 +27,54 @@ export const fakeUser: ExternalUser = Object.freeze({
...
+ *
+ * The `componentInstance` parameter is a React component instance.
+ * Example: <MyComponent {...props} />
+ *
+ * The `targetComponent` parameter is the React class (or function) that
+ * you want to retrieve from the component tree.
+ */
+type ShallowUntilTargetParams = {
+  componentInstance: React.ReactElement<any>;
+  targetComponent: any;
```
Comment (willdurand) [inline src/test-helpers.tsx]: Can we do better than `any` here?

File: src/components/App/index.spec.tsx (hunk 5 above)
Comment (bobsilverberg) [reply src/components/App/index.spec.tsx]: If this works, does that mean we aren't blocked on upgrading to react-redux v6?

Comment (willdurand) [reply src/components/App/index.spec.tsx]: We're still blocked by Enzyme 😞

File: src/test-helpers.tsx (hunk 6 above)
Comment (bobsilverberg) [reply src/test-helpers.tsx]: I tried a bunch of different things and couldn't seem to figure out what would work. I've now found that `Function` works. I'm not sure if that is 100% accurate, but it's definitely better than `any`.

Comment (willdurand) [reply src/test-helpers.tsx]: mmh, looking at some docs, I'd suggest `JSXElementConstructor<any>` (`any` refers to the props).

**Comment by willdurand** [comment]
This needs to be rebased now :)

**Comment by bobsilverberg** [comment]
> This needs to be rebased now :)

Hooray! 🎉

File: src/components/App/index.spec.tsx [hunk 7]
```diff
@@ -23,22 +25,15 @@ describe(__filename, () => {
   }: RenderParams = {}) => {
     const contextWithRouter = createContextWithFakeRouter();
     const context = {
-      ...contextWithRouter,
```
Comment (bobsilverberg) [inline src/components/App/index.spec.tsx]: @willdurand please take a look at this. I didn't understand why you were creating an object called `context` that only contains a single property called `context`, so I removed that. It could be that there might be something other than the `context` property in the object returned by `createContextWithFakeRouter()` at some point, but for now there is not, so the extra wrapping didn't really make sense to me.

Comment (willdurand) [reply src/components/App/index.spec.tsx]: `createContextWithFakeRouter()` is supposed to return:

```
{
  context: {},
  childContextTypes: {}
}
```

Comment (bobsilverberg) [reply src/components/App/index.spec.tsx]: Thanks @willdurand, I've put it back the way it was.

File: src/components/App/index.spec.tsx [hunk 8]
```diff
@@ -30,15 +32,11 @@ describe(__filename, () => {
...
-    const root = shallow(<App authToken={authToken} />, context)
-      // withRouter HOC
-      .shallow(context)
-      // connect HOC
-      .shallow(context)
-      // base component
-      .shallow();
+    const root = shallowUntilTarget({
+      componentInstance: <App authToken={authToken} />,
+      targetComponent: AppBase,
```
Comment (willdurand) [inline src/components/App/index.spec.tsx]: I would prefer to keep the old signature, we don't need object properties for the instance and the target component.

File: src/test-helpers.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+describe(__filename, () => {
+  describe('shallowUntilTarget', () => {
+    const ExampleBase = () => {
+      return <div>Example component</div>;
+    };
+
+    const wrapper = () => {
+      return function Wrapper(WrappedComponent: any) {
+        return function InnerWrapper(props: object) {
```
Comment (willdurand) [inline src/test-helpers.spec.tsx]: arrow function?

File: src/test-helpers.tsx [hunk 10]
```diff
@@ -80,3 +81,54 @@ export const createContextWithFakeRouter = ({
...
+    maxTries?: number;
+    shallowOptions?: object;
+    _shallow?: typeof shallow;
+  };
+};
+
+export const shallowUntilTarget = ({
+  componentInstance,
+  targetComponent,
+  options: { maxTries = 10, shallowOptions = {}, _shallow = shallow } = {},
```
Comment (willdurand) [inline src/test-helpers.tsx]: We don't want to assign an empty object here because that does not make much sense. It will fail if we don't specify some arguments, so I'd like to see the old signature (only the 3rd parameter is an object).

File: src/test-helpers.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+        render() {
+          return <div>example of using lifecycle methods</div>;
+        }
+      }
+
+      const Example = compose(wrapper())(LifecyleExample);
+
+      const root = shallowUntilTarget(<Example />, LifecyleExample, {
+        shallowOptions: { lifecycleExperimental: true },
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/test-helpers.spec.tsx]: You're right, that test is no longer valid. I changed it to use the `disableLifecycleMethods` option and I believe it is now valid once again.

File: src/test-helpers.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        _shallow: shallowStub,
+      });
+
+      expect(shallowStub).toHaveBeenCalledWith(instance, shallowOptions);
+    });
+
+    it('lets you pass options to the final shallow()', () => {
+      const componentDidUpdate = jest.fn();
+
+      // eslint-disable-next-line react/no-multi-comp
```
Comment (kumar303) [inline src/test-helpers.spec.tsx]: We should disable `react/no-multi-comp` in all test files. It's perfectly reasonable to put multiple components in test files.

Comment (willdurand) [reply src/test-helpers.spec.tsx]: I just filed: https://github.com/mozilla/addons-code-manager/issues/154
